### ISSUE TITLE
Normalize the database

### ISSUE THREAD

**Issue by jonasraoni** [pkp/pkp-lib#7315]
**Describe the problem you would like to solve**
Normalize the database in order to have better expectations (e.g. ensure a record is always identifiable by a single ID), improve the performance, decrease the amount of rows and also improve the data integrity.

**Describe the solution you'd like**
1. Add an `ID` field to all tables and discard composite primary keys.

2. Move non-localized fields from the `*_settings` tables to their origin table (e.g. `user_settings` > `user`), I think we have few cases, I just remember of seeing an `issueId` somewhere.

3. Move dynamic localized fields from the `*_settings` tables to a better typed table
- The moved field will be strongly typed instead of just a string, which will speed up queries against it
- Simplify the queries (instead of having a join for each field or an extra query returning all the settings, a single join will make all the localized fields available)
- Allow for customized indexes in specific fields
- Easier to upgrade (renaming a field vs updating the `setting_name` for N rows)
- Less prone for leaving orphan data behind
- The `*_setting` table might be left (reserved for plugins) or removed  altogether (plugins will deal with their data directly).

**Conversion process for the item #3**

Taking the `user_settings` as an example, where each user have at least two rows in the `user_settings` (one for the `givenName` and another for the `familyName`).

Create a `user_locale` table (`locale varchar, given_name varchar, family_name varchar, ...`), with all the localized fields, this way we'll have 1 record for each locale + user.
Update the queries to consider the new structure
Create migration scripts

**Extra**
If we're going to keep the `*_settings` for something else, perhaps we can use the JSON aggregate functions to return the data instead of having it loaded in a loop.

**Who is asking for this feature?**
Developer

**Comment by asmecher** [comment]
> 2. Move non-localized fields from the *_settings tables to their origin table (e.g. user_settings > user), I think we have few cases, I just remember of seeing an issueId somewhere.

I think you're referring to the `issueId` setting in `publication_settings`. Yes, this should definitely be normalized.

For the rest, I don't have a problem with what you're proposing per se, but it would be a pretty significant development effort. If it turns out that we're hitting e.g. a performance limit, then I could see prioritizing it -- but we'd need to establish that first.

**Comment by jonasraoni** [comment]
Yeah, it was the `issueId`! I understand it's tough and will introduce several breaking changes, I just wanted to leave the idea saved somewhere.

**Comment by NateWr** [comment]
I like the idea but my big concern with normalizing the `_settings` tables is that we need to continue to have an extensible data structure. If we normalize some fields (eg - `given_name`/`family_name`), we won't be able to get rid of the `_settings` table because plugins might make use of it to attach additional data.

Because of this, normalizing the core localized fields might lead to more complex SQL queries, because we would then have to join across an additional table (`users`, `user_locale` and `user_settings`).

I think probably our biggest performance issue is around searching/filtering. I wonder if it makes sense -- where we identify a performance issue -- to create a normalized table specifically for searching/filtering. That could really inflate the size of our DB, though.

**Comment by jonasraoni** [comment]
Yeah! I was updating a plugin and [saw this feature being used](https://github.com/ulsdevteam/pkp-betterPassword/blob/8364a4c2e3d2b093fc258790a7bb0a0b244f8004/BetterPasswordPlugin.inc.php#L100-L106) =]

- I'm totally ok to switch an extra join by data integrity! It's also easier to see what fields are available in the database, with the `settings` it's not possible to know it until some data is written or someone inspect the code.

- About removing the settings: My understanding is that a plugin might have a data structure that may not fit well into a setting table, so I'd be inclined to, in the long run, let them manage their data in a separated place and just offer a way to extend the queries (to return data in a single query)... But this should be discussed in another issue, for this one, I'd stick to the data normalization.

ps: I'll avoid discussing the real performance issues here, but I'm open for receiving pointers/data 😁
